File: app/src/main/java/dev/ragnarok/fenrir/api/impl/Networker.java
package dev.ragnarok.fenrir.api.impl;

import dev.ragnarok.fenrir.api.IOtherVkRetrofitProvider;
import dev.ragnarok.fenrir.api.IUploadRetrofitProvider;
import dev.ragnarok.fenrir.api.IVkRetrofitProvider;
import dev.ragnarok.fenrir.api.OtherVkRetrofitProvider;
import dev.ragnarok.fenrir.api.UploadRetrofitProvider;
import dev.ragnarok.fenrir.api.VkMethodHttpClientFactory;
import dev.ragnarok.fenrir.api.VkRetrofitProvider;
import dev.ragnarok.fenrir.api.interfaces.IAccountApis;
import dev.ragnarok.fenrir.api.interfaces.IAuthApi;
import dev.ragnarok.fenrir.api.interfaces.ILocalServerApi;
import dev.ragnarok.fenrir.api.interfaces.ILongpollApi;
import dev.ragnarok.fenrir.api.interfaces.INetworker;
import dev.ragnarok.fenrir.api.interfaces.IUploadApi;
import dev.ragnarok.fenrir.api.services.IAuthService;
import dev.ragnarok.fenrir.api.services.ILocalServerService;
import dev.ragnarok.fenrir.settings.IProxySettings;

public class Networker implements INetworker {

    private final IOtherVkRetrofitProvider otherVkRetrofitProvider;
    private final IVkRetrofitProvider vkRetrofitProvider;
    private final IUploadRetrofitProvider uploadRetrofitProvider;

    public Networker(IProxySettings settings) {
        otherVkRetrofitProvider = new OtherVkRetrofitProvider(settings);
        vkRetrofitProvider = new VkRetrofitProvider(settings, new VkMethodHttpClientFactory());
        uploadRetrofitProvider = new UploadRetrofitProvider(settings);
    }

    @Override
    public IAccountApis vkDefault(int accountId) {
        return VkApies.get(accountId, vkRetrofitProvider);
    }

    @Override
    public IAccountApis vkManual(int accountId, String accessToken) {
        return VkApies.create(accountId, accessToken, vkRetrofitProvider);
    }

    @Override
    public IAuthApi vkDirectAuth() {
        return new AuthApi(() -> otherVkRetrofitProvider.provideAuthRetrofit().map(wrapper -> wrapper.create(IAuthService.class)));
    }

    @Override
    public IAuthApi vkAuth() {
        return new AuthApi(() -> otherVkRetrofitProvider.provideAuthServiceRetrofit().map(wrapper -> wrapper.create(IAuthService.class)));
    }

    @Override
    public ILocalServerApi localServerApi() {
        return new LocalServerApi(() -> otherVkRetrofitProvider.provideLocalServerRetrofit().map(wrapper -> wrapper.create(ILocalServerService.class)));
    }

    @Override
    public ILongpollApi longpoll() {
        return new LongpollApi(otherVkRetrofitProvider);
    }

    @Override
    public IUploadApi uploads() {
        return new UploadApi(uploadRetrofitProvider);
    }
}

File: app/src/main/kotlin/dev/ragnarok/fenrir/api/impl/Networker.kt
package dev.ragnarok.fenrir.api.impl

import dev.ragnarok.fenrir.api.*
import dev.ragnarok.fenrir.api.interfaces.*
import dev.ragnarok.fenrir.api.services.IAuthService
import dev.ragnarok.fenrir.api.services.ILocalServerService
import dev.ragnarok.fenrir.settings.IProxySettings
import io.reactivex.rxjava3.core.Single

class Networker(settings: IProxySettings) : INetworker {
    private val otherVkRetrofitProvider: IOtherVkRetrofitProvider
    private val vkRetrofitProvider: IVkRetrofitProvider
    private val uploadRetrofitProvider: IUploadRetrofitProvider
    override fun vkDefault(accountId: Int): IAccountApis {
        return VkApies[accountId, vkRetrofitProvider]
    }

    override fun vkManual(accountId: Int, accessToken: String): IAccountApis {
        return VkApies.create(accountId, accessToken, vkRetrofitProvider)
    }

    override fun vkDirectAuth(): IAuthApi {
        return AuthApi(object : IDirectLoginSeviceProvider {
            override fun provideAuthService(): Single<IAuthService> {
                return otherVkRetrofitProvider.provideAuthRetrofit()
                    .map { wrapper ->
                        wrapper.create(
                            IAuthService::class.java
                        )
                    }
            }
        })
    }

    override fun vkAuth(): IAuthApi {
        return AuthApi(object : IDirectLoginSeviceProvider {
            override fun provideAuthService(): Single<IAuthService> {
                return otherVkRetrofitProvider.provideAuthServiceRetrofit()
                    .map { wrapper ->
                        wrapper.create(
                            IAuthService::class.java
                        )
                    }
            }
        })
    }

    override fun localServerApi(): ILocalServerApi {
        return LocalServerApi(object : ILocalServerServiceProvider {
            override fun provideLocalServerService(): Single<ILocalServerService> {
                return otherVkRetrofitProvider.provideLocalServerRetrofit()
                    .map { wrapper ->
                        wrapper.create(
                            ILocalServerService::class.java
                        )
                    }
            }
        })
    }

    override fun longpoll(): ILongpollApi {
        return LongpollApi(otherVkRetrofitProvider)
    }

    override fun uploads(): IUploadApi {
        return UploadApi(uploadRetrofitProvider)
    }

    init {
        otherVkRetrofitProvider = OtherVkRetrofitProvider(settings)
        vkRetrofitProvider = VkRetrofitProvider(settings, VkMethodHttpClientFactory())
        uploadRetrofitProvider = UploadRetrofitProvider(settings)
    }
}